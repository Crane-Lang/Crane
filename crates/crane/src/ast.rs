use smol_str::SmolStr;
use thin_vec::ThinVec;

/// An identifier.
#[derive(Debug, Clone)]
pub struct Ident(pub SmolStr);

impl std::fmt::Display for Ident {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// The kind of a Crane expression.
#[derive(Debug, Clone)]
pub enum ExprKind {
    /// A reference to a variable.
    Variable { name: Ident },

    /// A function call.
    Call {
        fun: Box<Expr>,
        args: ThinVec<Box<Expr>>,
    },
}

/// A Crane expression.
#[derive(Debug, Clone)]
pub struct Expr {
    pub kind: ExprKind,
}

/// The kind of a Crane statement.
#[derive(Debug, Clone)]
pub enum StmtKind {
    /// A function declaration (`fn`).
    Fn(Box<Fn>),

    /// An expression.
    Expr(Expr),
}

/// A function definition.
#[derive(Debug, Clone)]
pub struct Fn {
    pub name: SmolStr,
    pub body: ThinVec<Stmt>,
}

/// A Crane statement.
#[derive(Debug, Clone)]
pub struct Stmt {
    pub kind: StmtKind,
}
