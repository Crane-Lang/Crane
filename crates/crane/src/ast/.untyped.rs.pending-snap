{"run_id":"1687112702-689890000","line":225,"new":{"module_name":"crane__ast__untyped__tests","snapshot_name":"ast_node_sizes-4","metadata":{"source":"crates/crane/src/ast/untyped.rs","assertion_line":225,"expression":"size_of::<Item>().to_string()"},"snapshot":"64"},"old":{"module_name":"crane__ast__untyped__tests","metadata":{},"snapshot":"56"}}
{"run_id":"1687112777-693774000","line":222,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":223,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":224,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":225,"new":{"module_name":"crane__ast__untyped__tests","snapshot_name":"ast_node_sizes-4","metadata":{"source":"crates/crane/src/ast/untyped.rs","assertion_line":225,"expression":"size_of::<Item>().to_string()"},"snapshot":"64"},"old":{"module_name":"crane__ast__untyped__tests","metadata":{},"snapshot":"56"}}
{"run_id":"1687112794-630697000","line":222,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":223,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":224,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":225,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":226,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":227,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":228,"new":null,"old":null}
