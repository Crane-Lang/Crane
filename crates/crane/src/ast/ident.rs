use smol_str::SmolStr;

// TODO: Add span to `Ident`.
/// An identifier.
#[derive(Debug, PartialEq, Eq, Hash, Clone)]
pub struct Ident(pub SmolStr);

impl std::fmt::Display for Ident {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
