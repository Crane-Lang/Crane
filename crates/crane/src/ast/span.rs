use serde::{Deserialize, Serialize};

/// A dummy [`Span`].
///
/// All of the positions are set to `0`.
pub const DUMMY_SPAN: Span = Span { start: 0, end: 0 };

/// A span.
#[derive(PartialEq, Eq, Hash, Clone, Copy, Serialize, Deserialize)]
pub struct Span {
    /// The start offset of the span (inclusive).
    pub start: usize,

    /// The end offset of the span (exclusive).
    pub end: usize,
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    /// Returns a new [`Span`] that encloses `self and `end`.
    pub fn to(self, end: Span) -> Self {
        Self::new(
            std::cmp::min(self.start, end.start),
            std::cmp::max(self.end, end.end),
        )
    }
}

impl std::fmt::Debug for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}..{:?}", self.start, self.end)
    }
}

impl std::fmt::Display for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}..{}", self.start, self.end)
    }
}

impl From<std::ops::Range<usize>> for Span {
    fn from(value: std::ops::Range<usize>) -> Self {
        Self {
            start: value.start,
            end: value.end,
        }
    }
}

impl From<Span> for std::ops::Range<usize> {
    fn from(value: Span) -> Self {
        value.start..value.end
    }
}

impl ariadne::Span for Span {
    type SourceId = ();

    fn source(&self) -> &Self::SourceId {
        &()
    }

    fn start(&self) -> usize {
        self.start
    }

    fn end(&self) -> usize {
        self.end
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn span_to_works_forwards() {
        let start = Span::new(0, 3);
        let end = Span::new(4, 6);

        assert_eq!(start.to(end), Span::new(0, 6))
    }

    #[test]
    fn span_to_works_backwards() {
        let start = Span::new(3, 7);
        let end = Span::new(0, 2);

        assert_eq!(start.to(end), Span::new(0, 7))
    }
}
