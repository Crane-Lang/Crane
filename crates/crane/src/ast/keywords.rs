use smol_str::SmolStr;

use crate::ast::{Ident, DUMMY_SPAN};

pub const FN: Ident = Ident {
    name: SmolStr::new_inline("fn"),
    span: DUMMY_SPAN,
};

pub const LET: Ident = Ident {
    name: SmolStr::new_inline("let"),
    span: DUMMY_SPAN,
};

pub const MOD: Ident = Ident {
    name: SmolStr::new_inline("mod"),
    span: DUMMY_SPAN,
};

pub const PUB: Ident = Ident {
    name: SmolStr::new_inline("pub"),
    span: DUMMY_SPAN,
};

pub const STRUCT: Ident = Ident {
    name: SmolStr::new_inline("struct"),
    span: DUMMY_SPAN,
};

pub const UNION: Ident = Ident {
    name: SmolStr::new_inline("union"),
    span: DUMMY_SPAN,
};
