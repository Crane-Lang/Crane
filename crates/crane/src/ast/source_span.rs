use crate::ast::Span;

/// A [`Span`] with the source from where it originated.
pub struct SourceSpan {
    pub source: String,
    pub span: Span,
}

impl From<(&str, Span)> for SourceSpan {
    fn from(value: (&str, Span)) -> Self {
        Self {
            source: value.0.parse().unwrap(),
            span: value.1,
        }
    }
}

impl ariadne::Span for SourceSpan {
    type SourceId = String;

    fn source(&self) -> &Self::SourceId {
        &self.source
    }

    fn start(&self) -> usize {
        self.span.start
    }

    fn end(&self) -> usize {
        self.span.end
    }
}
