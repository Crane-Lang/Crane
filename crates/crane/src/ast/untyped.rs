use smol_str::SmolStr;
use thin_vec::ThinVec;

use crate::ast::{Ident, Span};

/// The kind of an [`Expr`].
#[derive(Debug, Clone)]
pub enum ExprKind {
    /// A literal.
    Literal(Literal),

    /// A reference to a variable.
    Variable { name: Ident },

    /// A function call.
    Call {
        fun: Box<Expr>,
        args: ThinVec<Box<Expr>>,
    },
}

/// The kind of a [`Literal`].
#[derive(Debug, Clone)]
pub enum LiteralKind {
    String,
}

/// A literal.
#[derive(Debug, Clone)]
pub struct Literal {
    pub kind: LiteralKind,
    pub value: SmolStr,
}

/// An expression.
#[derive(Debug, Clone)]
pub struct Expr {
    pub kind: ExprKind,
    pub span: Span,
}

/// The kind of a [`Stmt`].
#[derive(Debug, Clone)]
pub enum StmtKind {
    /// An item.
    Item(Item),

    /// An expression.
    Expr(Expr),
}

/// A function definition.
#[derive(Debug, Clone)]
pub struct Fn {
    pub params: ThinVec<FnParam>,
    pub body: ThinVec<Stmt>,
}

/// A parameter to a [`Fn`].
#[derive(Debug, Clone)]
pub struct FnParam {
    pub name: Ident,
    pub span: Span,
}

/// A statement.
#[derive(Debug, Clone)]
pub struct Stmt {
    pub kind: StmtKind,
    pub span: Span,
}

/// The kind of an [`Item`].
#[derive(Debug, Clone)]
pub enum ItemKind {
    /// A function declaration (`fn`).
    Fn(Box<Fn>),
}

/// An item in a [`Module`].
#[derive(Debug, Clone)]
pub struct Item {
    pub kind: ItemKind,
    pub name: Ident,
}

/// A module.
#[derive(Debug, Clone)]
pub struct Module {
    pub items: ThinVec<Item>,
}
