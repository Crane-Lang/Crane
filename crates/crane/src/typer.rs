mod environment;
mod error;
mod r#type;

pub use error::*;
pub use r#type::*;

use std::collections::HashMap;
use std::sync::Arc;

use smol_str::SmolStr;
use thin_vec::{thin_vec, ThinVec};

use crate::ast::{
    Expr, ExprKind, Fn, FnParam, Ident, InlineModuleDecl, Item, ItemKind, Literal, LiteralKind,
    Local, LocalKind, Module, ModuleDecl, Package, Span, Stmt, StmtKind, StructDecl, TyExpr,
    TyExprKind, TyFieldDecl, TyFn, TyFnParam, TyIntegerLiteral, TyItem, TyItemKind, TyLiteral,
    TyLiteralKind, TyLocal, TyLocalKind, TyModule, TyPackage, TyPath, TyPathSegment, TyStmt,
    TyStmtKind, TyStructDecl, TyUint, TyUnionDecl, TyVariant, TyVariantData, UnionDecl,
    VariantData, DUMMY_SPAN,
};

fn ty_to_string(ty: &Type) -> String {
    match ty {
        Type::UserDefined { module, name } => {
            format!("{}::{}", module, name)
        }
        Type::Fn { args, return_ty } => format!(
            "Fn({}) -> {}",
            args.iter()
                .map(|ty| ty_to_string(ty))
                .collect::<Vec<_>>()
                .join(", "),
            ty_to_string(return_ty)
        ),
    }
}

pub type TypeCheckResult<T> = Result<T, TypeError>;

pub struct Typer {
    module_functions: HashMap<TyPath, (ThinVec<TyFnParam>, Arc<Type>)>,
    scopes: Vec<HashMap<TyPath, Arc<Type>>>,
}

impl Typer {
    pub fn new() -> Self {
        Self {
            module_functions: HashMap::new(),
            scopes: Vec::new(),
        }
    }

    pub fn type_check_package(&mut self, package: Package) -> TypeCheckResult<TyPackage> {
        let mut typed_modules = ThinVec::new();

        for module in package.modules {
            typed_modules.push(self.type_check_module(module)?);
        }

        Ok(TyPackage {
            modules: typed_modules,
        })
    }

    pub fn type_check_module(&mut self, module: Module) -> TypeCheckResult<TyModule> {
        let unit_ty = Arc::new(Type::UserDefined {
            module: SmolStr::new_inline("std::prelude"),
            name: SmolStr::new_inline("()"),
        });

        let string_ty = Arc::new(Type::UserDefined {
            module: SmolStr::new_inline("std::prelude"),
            name: SmolStr::new_inline("String"),
        });

        let uint64_ty = Arc::new(Type::UserDefined {
            module: SmolStr::new_inline("std::prelude"),
            name: SmolStr::new_inline("Uint64"),
        });

        // HACK: Register the functions from `std`.
        self.register_function(
            TyPath {
                segments: thin_vec![
                    TyPathSegment {
                        ident: Ident {
                            name: "std".into(),
                            span: DUMMY_SPAN,
                        }
                    },
                    TyPathSegment {
                        ident: Ident {
                            name: "io".into(),
                            span: DUMMY_SPAN,
                        }
                    },
                    TyPathSegment {
                        ident: Ident {
                            name: "print".into(),
                            span: DUMMY_SPAN,
                        }
                    }
                ],
                span: DUMMY_SPAN,
            },
            thin_vec![TyFnParam {
                name: Ident {
                    name: "value".into(),
                    span: DUMMY_SPAN
                },
                ty: string_ty.clone(),
                span: DUMMY_SPAN
            }],
            unit_ty.clone(),
        );
        self.register_function(
            TyPath {
                segments: thin_vec![
                    TyPathSegment {
                        ident: Ident {
                            name: "std".into(),
                            span: DUMMY_SPAN,
                        }
                    },
                    TyPathSegment {
                        ident: Ident {
                            name: "io".into(),
                            span: DUMMY_SPAN,
                        }
                    },
                    TyPathSegment {
                        ident: Ident {
                            name: "println".into(),
                            span: DUMMY_SPAN,
                        }
                    }
                ],
                span: DUMMY_SPAN,
            },
            thin_vec![TyFnParam {
                name: Ident {
                    name: "value".into(),
                    span: DUMMY_SPAN
                },
                ty: string_ty.clone(),
                span: DUMMY_SPAN
            }],
            unit_ty.clone(),
        );
        self.register_function(
            TyPath {
                segments: thin_vec![
                    TyPathSegment {
                        ident: Ident {
                            name: "std".into(),
                            span: DUMMY_SPAN,
                        }
                    },
                    TyPathSegment {
                        ident: Ident {
                            name: "int".into(),
                            span: DUMMY_SPAN,
                        }
                    },
                    TyPathSegment {
                        ident: Ident {
                            name: "int_add".into(),
                            span: DUMMY_SPAN,
                        }
                    }
                ],
                span: DUMMY_SPAN,
            },
            thin_vec![
                TyFnParam {
                    name: Ident {
                        name: "a".into(),
                        span: DUMMY_SPAN
                    },
                    ty: uint64_ty.clone(),
                    span: DUMMY_SPAN
                },
                TyFnParam {
                    name: Ident {
                        name: "b".into(),
                        span: DUMMY_SPAN
                    },
                    ty: uint64_ty.clone(),
                    span: DUMMY_SPAN
                }
            ],
            uint64_ty.clone(),
        );
        self.register_function(
            TyPath {
                segments: thin_vec![
                    TyPathSegment {
                        ident: Ident {
                            name: "std".into(),
                            span: DUMMY_SPAN,
                        }
                    },
                    TyPathSegment {
                        ident: Ident {
                            name: "int".into(),
                            span: DUMMY_SPAN,
                        }
                    },
                    TyPathSegment {
                        ident: Ident {
                            name: "int_to_string".into(),
                            span: DUMMY_SPAN,
                        }
                    }
                ],
                span: DUMMY_SPAN,
            },
            thin_vec![TyFnParam {
                name: Ident {
                    name: "value".into(),
                    span: DUMMY_SPAN
                },
                ty: uint64_ty,
                span: DUMMY_SPAN
            }],
            string_ty,
        );

        for item in &module.items {
            match item.kind {
                ItemKind::Use(_) => {}
                ItemKind::Fn(ref fun) => {
                    let params = self.infer_function_params(&fun.params)?;

                    let return_ty = fun
                        .return_ty
                        .as_ref()
                        .map(|return_ty| {
                            Arc::new(Type::UserDefined {
                                module: "std::prelude".into(),
                                name: return_ty.to_string().into(),
                            })
                        })
                        .unwrap_or(unit_ty.clone());

                    self.register_function(
                        TyPath {
                            segments: thin_vec![TyPathSegment {
                                ident: item.name.clone()
                            }],
                            span: item.name.span,
                        },
                        params.clone(),
                        return_ty,
                    );
                }
                ItemKind::Struct(_) => {}
                ItemKind::Union(_) => {}
                ItemKind::Module(ref module_decl) => match *module_decl.clone() {
                    ModuleDecl::Loaded(module, InlineModuleDecl::Yes) => {
                        self.type_check_module(module.clone())?;
                    }
                    ModuleDecl::Loaded(_, InlineModuleDecl::No) => todo!(),
                    ModuleDecl::Unloaded => todo!(),
                },
            }
        }

        let mut typed_items = ThinVec::new();

        for item in module.items {
            typed_items.push(self.infer_item(item)?);
        }

        Ok(TyModule { items: typed_items })
    }

    fn register_function(
        &mut self,
        path: TyPath,
        params: ThinVec<TyFnParam>,
        return_ty: Arc<Type>,
    ) {
        self.module_functions.insert(path, (params, return_ty));
    }

    fn ensure_function_exists(
        &self,
        path: &TyPath,
    ) -> TypeCheckResult<(&ThinVec<TyFnParam>, Arc<Type>)> {
        if let Some((params, return_ty)) = self.module_functions.get(path) {
            return Ok((params, return_ty.clone()));
        }

        Err(TypeError {
            kind: TypeErrorKind::UnknownFunction(path.clone()),
            span: path.span,
        })
    }

    fn infer_item(&mut self, item: Item) -> TypeCheckResult<TyItem> {
        match item.kind {
            ItemKind::Use(_) => Ok(TyItem {
                kind: TyItemKind::Use,
                name: item.name,
            }),
            ItemKind::Fn(fun) => Ok(TyItem {
                kind: TyItemKind::Fn(Box::new(self.infer_function(&item.name, *fun)?)),
                name: item.name,
            }),
            ItemKind::Struct(struct_decl) => Ok(TyItem {
                kind: TyItemKind::Struct(self.infer_struct_decl(&struct_decl)?),
                name: item.name,
            }),
            ItemKind::Union(union_decl) => Ok(TyItem {
                kind: TyItemKind::Union(self.infer_union_decl(&union_decl)?),
                name: item.name,
            }),
            ItemKind::Module(module_decl) => Ok(TyItem {
                kind: TyItemKind::Module(self.infer_module_decl(&module_decl)?),
                name: item.name,
            }),
        }
    }

    fn infer_function(&mut self, name: &Ident, fun: Fn) -> TypeCheckResult<TyFn> {
        let path = TyPath {
            segments: thin_vec![TyPathSegment {
                ident: name.clone()
            }],
            span: name.span,
        };

        let (_, return_ty) = self.ensure_function_exists(&path)?;

        let params = self.infer_function_params(&fun.params)?;

        self.scopes
            .push(HashMap::from_iter(params.clone().into_iter().map(
                |param| {
                    (
                        TyPath {
                            segments: thin_vec![TyPathSegment { ident: param.name }],
                            span: param.span,
                        },
                        param.ty,
                    )
                },
            )));

        let body = fun
            .body
            .into_iter()
            .map(|stmt| self.infer_stmt(stmt))
            .collect::<Result<ThinVec<_>, _>>()?;

        if let Some(last_stmt) = body.last() {
            let ty = match &last_stmt.kind {
                TyStmtKind::Local(_) => todo!(),
                TyStmtKind::Expr(expr) => &expr.ty,
                TyStmtKind::Item(_) => todo!(),
            };

            if *ty != return_ty {
                return Err(TypeError {
                    kind: TypeErrorKind::Error(format!(
                        "Expected `{name}` to return {} but got {}",
                        ty_to_string(&return_ty),
                        ty_to_string(&ty)
                    )),
                    span: last_stmt.span,
                });
            }
        }

        let ty_fn = TyFn {
            params,
            return_ty,
            body,
        };

        self.scopes.pop();

        Ok(ty_fn)
    }

    fn infer_function_params(
        &self,
        params: &ThinVec<FnParam>,
    ) -> TypeCheckResult<ThinVec<TyFnParam>> {
        Ok(params
            .iter()
            .map(|param| TyFnParam {
                name: param.name.clone(),
                ty: Arc::new(Type::UserDefined {
                    module: "std::prelude".into(),
                    name: param.ty.to_string().into(),
                }),
                span: param.span,
            })
            .collect::<ThinVec<_>>())
    }

    fn infer_struct_decl(&self, struct_decl: &StructDecl) -> TypeCheckResult<TyStructDecl> {
        Ok(TyStructDecl(self.infer_variant_data(&struct_decl.0)?))
    }

    fn infer_union_decl(&self, union_decl: &UnionDecl) -> TypeCheckResult<TyUnionDecl> {
        let ty_variants = union_decl
            .variants
            .iter()
            .map(|variant| {
                Ok(TyVariant {
                    name: variant.name.clone(),
                    data: self.infer_variant_data(&variant.data)?,
                    span: variant.span,
                })
            })
            .collect::<Result<ThinVec<_>, _>>()?;

        Ok(TyUnionDecl {
            variants: ty_variants,
        })
    }

    fn infer_module_decl(&mut self, module_decl: &ModuleDecl) -> TypeCheckResult<TyModule> {
        match &module_decl {
            ModuleDecl::Loaded(module, InlineModuleDecl::Yes) => Ok(TyModule {
                items: module
                    .items
                    .iter()
                    .map(|item| self.infer_item(item.clone()))
                    .collect::<Result<ThinVec<_>, _>>()?,
            }),
            ModuleDecl::Loaded(_, InlineModuleDecl::No) => Err(TypeError {
                kind: TypeErrorKind::Error("Only inline modules are supported.".into()),
                span: DUMMY_SPAN,
            }),
            ModuleDecl::Unloaded => Err(TypeError {
                kind: TypeErrorKind::Error("Only inline modules are supported.".into()),
                span: DUMMY_SPAN,
            }),
        }
    }

    fn infer_variant_data(&self, variant_data: &VariantData) -> TypeCheckResult<TyVariantData> {
        Ok(match &variant_data {
            VariantData::Struct(fields) => TyVariantData::Struct(
                fields
                    .into_iter()
                    .map(|field| TyFieldDecl {
                        name: field.name.clone(),
                        ty: field.ty.clone(),
                        span: field.span,
                    })
                    .collect::<ThinVec<_>>(),
            ),
            VariantData::Tuple(_) => todo!(),
            VariantData::Unit => TyVariantData::Unit,
        })
    }

    fn infer_stmt(&mut self, stmt: Stmt) -> TypeCheckResult<TyStmt> {
        Ok(TyStmt {
            kind: match stmt.kind {
                StmtKind::Local(local) => TyStmtKind::Local(Box::new(self.infer_local(*local)?)),
                StmtKind::Expr(expr) => TyStmtKind::Expr(Box::new(self.infer_expr(*expr)?)),
                StmtKind::Item(_) => todo!(),
            },
            span: stmt.span,
        })
    }

    fn infer_local(&mut self, local: Local) -> TypeCheckResult<TyLocal> {
        let ty = match local.kind.init() {
            Some(init) => self.infer_expr(init.clone())?.ty,
            None => Arc::new(Type::UserDefined {
                module: "?".into(),
                name: "?".into(),
            }),
        };

        if let Some(scope) = self.scopes.last_mut() {
            scope.insert(
                TyPath {
                    segments: thin_vec![TyPathSegment {
                        ident: local.name.clone()
                    }],
                    span: local.span,
                },
                ty.clone(),
            );
        }

        Ok(TyLocal {
            kind: match local.kind {
                LocalKind::Decl => TyLocalKind::Decl,
                LocalKind::Init(init) => TyLocalKind::Init(Box::new(self.infer_expr(*init)?)),
            },
            name: local.name,
            ty: Some(ty),
            span: local.span,
        })
    }

    fn infer_expr(&self, expr: Expr) -> TypeCheckResult<TyExpr> {
        match expr.kind {
            ExprKind::Literal(literal) => match literal.kind {
                LiteralKind::String => self.infer_string(literal, expr.span),
                LiteralKind::Integer => self.infer_integer(literal, expr.span),
            },
            ExprKind::Variable(path) => {
                let path = TyPath {
                    segments: path
                        .segments
                        .into_iter()
                        .map(|segment| TyPathSegment {
                            ident: segment.ident,
                        })
                        .collect::<ThinVec<_>>(),
                    span: path.span,
                };

                let ty = self
                    .scopes
                    .last()
                    .and_then(|scope| scope.get(&path).cloned());

                Ok(TyExpr {
                    kind: TyExprKind::Variable(path),
                    ty: ty.unwrap_or_else(|| {
                        Arc::new(Type::UserDefined {
                            module: "?".into(),
                            name: "?".into(),
                        })
                    }),
                    span: expr.span,
                })
            }
            ExprKind::Call { fun, args } => {
                let callee = self.infer_expr(*fun.clone())?;

                let callee = match &callee.kind {
                    TyExprKind::Variable(path) => Ok(path),
                    _ => Err(TypeError {
                        kind: TypeErrorKind::Error("Not a function.".to_string()),
                        span: callee.span,
                    }),
                }?;

                let (callee_params, callee_return_ty) =
                    self.module_functions.get(callee).ok_or_else(|| TypeError {
                        kind: TypeErrorKind::Error(format!("Function `{callee}` not found.",)),
                        span: callee.span,
                    })?;

                let caller_args = args
                    .into_iter()
                    .map(|expr| self.infer_expr(*expr))
                    .map(|result| result.map(Box::new))
                    .collect::<Result<ThinVec<_>, _>>()?;

                let callee_arity = callee_params.len();
                let caller_arity = caller_args.len();

                if callee_arity != caller_arity {
                    return Err(TypeError {
                        kind: TypeErrorKind::Error(format!("`{callee}` was called with {caller_arity} arguments when it expected {callee_arity}")),
                        span: callee.span
                    });
                }

                for (param, arg) in callee_params.into_iter().zip(&caller_args) {
                    if param.ty != arg.ty {
                        return Err(TypeError {
                            kind: TypeErrorKind::Error(format!(
                                "Expected `{}` but received `{}`",
                                ty_to_string(&param.ty),
                                ty_to_string(&arg.ty)
                            )),
                            span: arg.span,
                        });
                    }
                }

                Ok(TyExpr {
                    kind: TyExprKind::Call {
                        fun: Box::new(self.infer_expr(*fun)?),
                        args: caller_args,
                    },
                    ty: callee_return_ty.clone(),
                    span: expr.span,
                })
            }
        }
    }

    fn infer_string(&self, literal: Literal, span: Span) -> TypeCheckResult<TyExpr> {
        Ok(TyExpr {
            kind: TyExprKind::Literal(TyLiteral {
                kind: TyLiteralKind::String(literal.value),
                span,
            }),
            span,
            ty: Arc::new(Type::UserDefined {
                module: "std::prelude".into(),
                name: "String".into(),
            }),
        })
    }

    fn infer_integer(&self, literal: Literal, span: Span) -> TypeCheckResult<TyExpr> {
        let value: u128 = literal.value.parse().expect("Failed to parse integer.");

        Ok(TyExpr {
            kind: TyExprKind::Literal(TyLiteral {
                kind: TyLiteralKind::Integer(TyIntegerLiteral::Unsigned(value, TyUint::Uint64)),
                span,
            }),
            span,
            ty: Arc::new(Type::UserDefined {
                module: "std::prelude".into(),
                name: "Uint64".into(),
            }),
        })
    }
}

#[cfg(test)]
mod tests {
    use crate::lexer::Lexer;
    use crate::parser::Parser;

    use super::*;

    #[test]
    pub fn test_typer() {
        insta::glob!("snapshot_inputs/*.crane", |path| {
            let source = std::fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&source);
            let parser = Parser::new(lexer);

            let items = parser.parse().unwrap();

            let module = Module { items };

            let mut typer = Typer::new();

            insta::assert_yaml_snapshot!(typer.type_check_module(module));
        })
    }
}
