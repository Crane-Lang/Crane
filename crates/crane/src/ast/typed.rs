use std::sync::Arc;

use serde::{Deserialize, Serialize};
use smol_str::SmolStr;
use thin_vec::ThinVec;

use crate::ast::{Ident, Span};
use crate::typer::Type;

/// The type of an unsigned integer.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TyUint {
    Uint64,
}

/// The type of an integer literal.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TyIntegerLiteral {
    Unsigned(u128, TyUint),
}

/// The kind of a [`TyLiteral`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TyLiteralKind {
    String(SmolStr),
    Integer(TyIntegerLiteral),
}

/// A typed literal.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyLiteral {
    pub kind: TyLiteralKind,
    pub span: Span,
}

/// The kind of a [`TyExpr`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TyExprKind {
    /// A literal.
    Literal(TyLiteral),

    /// A reference to a variable.
    Variable { name: Ident },

    /// A function call.
    Call {
        fun: Box<TyExpr>,
        args: ThinVec<Box<TyExpr>>,
    },
}

/// A typed expression.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyExpr {
    pub kind: TyExprKind,
    pub span: Span,
    pub ty: Arc<Type>,
}

/// The kind of a [`TyStmt`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TyStmtKind {
    /// An item.
    Item(TyItem),

    /// An expression.
    Expr(TyExpr),
}

/// A typed function definition.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyFn {
    pub params: ThinVec<TyFnParam>,
    pub return_ty: Arc<Type>,
    pub body: ThinVec<TyStmt>,
}

/// A parameter to a [`TyFn`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyFnParam {
    pub name: Ident,
    pub ty: Arc<Type>,
    pub span: Span,
}

/// A typed statement.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyStmt {
    pub kind: TyStmtKind,
    pub span: Span,
}

/// A field declaration in a `struct` or [`Variant`] of a `union`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyFieldDecl {
    pub name: Option<Ident>,
    pub ty: Ident,
    pub span: Span,
}

/// The data for a `struct` or [`Variant`] of a `union`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TyVariantData {
    /// A struct variant.
    Struct(ThinVec<TyFieldDecl>),

    /// A tuple variant.
    Tuple(ThinVec<TyFieldDecl>),

    /// A unit variant.
    Unit,
}

impl TyVariantData {
    /// Return the fields of this variant.
    pub fn fields(&self) -> &[TyFieldDecl] {
        match self {
            TyVariantData::Struct(fields) | TyVariantData::Tuple(fields) => fields,
            _ => &[],
        }
    }
}

/// A variant in a `union`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyVariant {
    pub name: Ident,
    pub data: TyVariantData,
    pub span: Span,
}

/// A `struct` declaration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyStructDecl(pub TyVariantData);

/// A `union` declaration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyUnionDecl {
    /// The variants of the `union`.
    pub variants: ThinVec<TyVariant>,
}

/// The kind of a [`TyItem`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum TyItemKind {
    // TODO: Remove `Box`?
    /// A function declaration (`fn`).
    Fn(Box<TyFn>),

    /// A struct declaration (`struct`).
    Struct(TyStructDecl),

    /// A union declaration (`union`).
    Union(TyUnionDecl),
}

/// An item in a [`TyModule`].
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyItem {
    pub kind: TyItemKind,
    pub name: Ident,
}

/// A module.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TyModule {
    pub items: ThinVec<TyItem>,
}

#[cfg(test)]
mod tests {
    #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
    use super::*;

    /// Tests the size of AST nodes to ensure they don't unintentionally get bigger.
    #[test]
    #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
    fn test_ast_node_sizes() {
        use std::mem::size_of;

        dbg!(size_of::<TyExpr>().to_string());
        dbg!(size_of::<TyExprKind>().to_string());
        dbg!(size_of::<TyFn>().to_string());
        dbg!(size_of::<TyItem>().to_string());
        dbg!(size_of::<TyItemKind>().to_string());
        dbg!(size_of::<TyStmt>().to_string());
        dbg!(size_of::<TyStmtKind>().to_string());

        // insta::assert_snapshot!(size_of::<TyExpr>().to_string(), @"80");
        // insta::assert_snapshot!(size_of::<TyExprKind>().to_string(), @"48");
        // insta::assert_snapshot!(size_of::<TyFn>().to_string(), @"24");
        // insta::assert_snapshot!(size_of::<TyItem>().to_string(), @"56");
        // insta::assert_snapshot!(size_of::<TyItemKind>().to_string(), @"16");
        // insta::assert_snapshot!(size_of::<TyStmt>().to_string(), @"96");
        // insta::assert_snapshot!(size_of::<TyStmtKind>().to_string(), @"80");
    }
}
