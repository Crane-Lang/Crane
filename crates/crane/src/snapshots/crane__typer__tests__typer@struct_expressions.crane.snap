---
source: crates/crane/src/typer.rs
expression: typer.type_check_package(package)
input_file: crates/crane/src/snapshot_inputs/struct_expressions.crane
---
Err:
  kind:
    Error: Struct expressions are not yet supported.
  span:
    start: 108
    end: 195

